### PR TITLE
feat: Prototype Module-Acceleration in Dynamo

### LINKED ISSUE #1894: ✨[Feature] Upgrade Prototype System for Module-Level Acceleration in Dynamo Path
- Add functionality to intercept, exclude from tracing, and replace module-size code blocks in FX code
- Add tutorial code for easy customization and inclusion of new operators

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Add support for excluding entire Torch modules from tracing in Dynamo using Torch custom operators
- Develop new dataclass to store required replacement functions and operators in a streamlined way
- Add new registry to store mapping between replacement operators and their corresponding dataclass
- Add detailed tutorial descriptions to sample module substitution with step-by-step detailed instructions for creating a new module substitution

Fixes #1894 

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -258,7 +258,7 @@ commands:
             name: Set up python environment
             command: |
               pip3 install --upgrade pip
-              pip3 install wheel setuptools
+              pip3 install wheel setuptools pyyaml
```
Comment (narendasan) [inline .circleci/config.yml:261]: Why do we need this dep? Is it something we need to add to setup.py / requirements.txt?

File: py/torch_tensorrt/dynamo/backend/lowering/_partition.py [hunk 2]
```diff
@@ -14,6 +15,11 @@
 
 logger = logging.getLogger(__name__)
 
+DEFAULT_SINGLE_NODE_PARTITIONS: Set[str] = set(
+    "torch.ops." + str(module.new_operator)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/lowering/_partition.py:19]: Do you know if theres a better type than string for this registry? Like is there a op type?

File: py/torch_tensorrt/dynamo/backend/lowering/_pre_aot_lowering.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    subgraph_insertion_fn: Callable[
+        [torch.fx.GraphModule, torch.nn.Module, torch.fx.Node], torch.fx.Node
+    ]
+
+
+# Dictionary mapping module to ModuleReplacement instance
+MODULE_SUBSTITUTION_REGISTRY: Dict[torch.nn.Module, ModuleReplacement] = dict()
+
+
+def module_substitution(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/lowering/_pre_aot_lowering.py:29]: This is sick 😄

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+from typing import Dict, Tuple
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:1]: It would be cool to have a sphinx tutorial on how to do this from an external user perspective. could be as easy as removing maxpool1d from the registry then walking through all the parts.

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    ...
+
+
+@maxpool1d.impl("cpu")
+@maxpool1d.impl("cuda")
+def maxpool1d_generic(
+    *args,
+    **kwargs,
+):
+    # Defines a converter implementation for AOT Autograd to use for shape analysis/propagation
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:28]: Just an implementation right not a converter?

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    return torch.nn.functional.max_pool1d(
+        *args,
+        **kwargs,
+    )
+
+
+@tensorrt_converter(torch.ops.tensorrt.maxpool1d.default)
+def aten_ops_maxpool1d(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:36]: For module substitution "in library" do we want to put the converter here? or do we want to put the converter in the registry with the rest?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:36]: For external users they'd probably put it here.

File: py/torch_tensorrt/dynamo/backend/test/utils.py [hunk 7]
```diff
@@ -31,6 +34,8 @@ def fx_dynamo_testing_backend(
         torch_executed_ops=torch_executed_ops,
     )
 
+    gm = pre_aot_module_replacement(gm)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/backend/test/utils.py:37]: Why is there a separate testing backend? Would we need to continue to make changes to this in step with the actual one?

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/test/utils.py:37]: The main reason for the separate testing backend is the argument `store_intermediate_graphs`. This is needed to track intermediate partitioned modules at different points in the compilation, to ensure decompositions, fusions, etc. are being utilized.

As changes are made to the main backend, yes, those changes would need to be reflected here, and in the `compile_module_testing` function below, however these are higher-level functions which do not change often.

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py (hunk 6 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:36]: I thought it could be cleaner to have the converter implementation here, so all of the code relating to that module and its replacement is centralized.

The requirement, however, is that for every new module replacement file, the user will have to add `from .my_module_replacement import *` to `module_substitutions/__init__.py` to ensure the registrations occur.

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py (hunk 4 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:1]: Agreed. I will add this, in addition to documentation on how to ensure all the relevant code is registered (`__init__.py` imports)

File: py/torch_tensorrt/dynamo/backend/lowering/_partition.py (hunk 2 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/lowering/_partition.py:19]: I will look into this more - there is the `torch._ops.OpOverload` type which could be a substitute here.

File: .circleci/config.yml (hunk 1 above)
Comment (gs-olive) [reply .circleci/config.yml:261]: I had noticed some CI failures when not using this - I think it may be necessary on account of one of the new imports, I will check.

File: py/torch_tensorrt/dynamo/backend/lowering/_pre_aot_lowering.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    # Function taking a containing graph, a submodule, and a 'call_module' node and returning
+    # a replacement node, with type 'call_function', or raising an Error if incompatibility is detected
+    # Note: subgraph_insertion_fn should NOT delete nodes or recompile the graph
+    subgraph_insertion_fn: Callable[
+        [torch.fx.GraphModule, torch.nn.Module, torch.fx.Node], torch.fx.Node
+    ]
+
+
+# Dictionary mapping module to ModuleReplacement instance
+MODULE_SUBSTITUTION_REGISTRY: Dict[torch.nn.Module, ModuleReplacement] = dict()
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/backend/lowering/_pre_aot_lowering.py:26]: Change to `Type[torch.nn.Module]`

File: py/torch_tensorrt/dynamo/backend/lowering/_partition.py (hunk 2 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/lowering/_partition.py:19]: After looking more into this, `str` is likely the best choice for this registry for now. The issue is that node targets can be a mix of functions and `torch._ops.OpOverload` objects. For example, `torch.ops.aten.add.Tensor` is an overload object representing a Tensor addition op, whereas the operator for `get` is an actual Python function. The unifying framework which can connect all of these types is the [`_get_qualified_name`](https://github.com/pytorch/pytorch/blob/17166c2511fa8b18378438b4283f2d2b4625356e/torch/fx/node.py#L79) function, which can handle all of these types, and returns a string. I have updated the implementation here to use that function.

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py (hunk 4 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:1]: To avoid clashing with or making dependencies on #1966 and #1967, I'll add more commenting to this file and then port it over to a formatted `.py` file once those PR are merged.

File: py/torch_tensorrt/dynamo/backend/lowering/_partition.py [hunk 9]
```diff
@@ -14,6 +15,11 @@
 
 logger = logging.getLogger(__name__)
 
+DEFAULT_SINGLE_NODE_PARTITIONS: Set[str] = set(
+    _get_qualified_name(module.new_operator)
+    for module in MODULE_SUBSTITUTION_REGISTRY.values()
+)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/backend/lowering/_partition.py:21]: Updated registry to use strings formatted in a streamlined/uniform way via [`get_qualified_name`](https://github.com/pytorch/pytorch/blob/17166c2511fa8b18378438b4283f2d2b4625356e/torch/fx/node.py#L79) as is already done in partitioning, and addressing the issues raised here: https://github.com/pytorch/TensorRT/pull/1921#discussion_r1212225354

File: py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+# preferable implementation order.
+
+
+# 1. The Placeholder
+#
+# Specify the schema and namespace of the operator, as well as a placeholder function
+# representing the schema. The schema should be in torch JIT syntax, indicating input and output
+# types. The namespace, such as tensorrt, will cause the op to be registered as torch.ops.tensorrt.your_op
+# Then, create a placeholder function with no operations, but having the same schema and naming as that
+# used in the decorator
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/backend/lowering/module_substitutions/maxpool1d.py:24]: Added detailed instructions for creating a module substitution

**Comment by gs-olive** [comment]
Closed in place of #1979
